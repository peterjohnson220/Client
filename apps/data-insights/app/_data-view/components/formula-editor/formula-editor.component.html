<div class="d-flex flex-column">
  <div class="formula-container mb-3">
    <div class="popup-anchor formula-suggestions">
      <span>{{beforeMarkerValues}}</span>
      <span>
        {{markerValue}}
        <kendo-dropdownlist #suggestionList
          [style.visibility]="showPopup ? 'visible' : 'hidden'"
          [value]="searchTerm"
          [data]="filteredSuggestions"
          [popupSettings]="popupSettings"
          [textField]="'Name'"
          [valueField]="'Value'"
          [valuePrimitive]="true"
          (valueChange)="handleSuggestionClicked($event)">
        </kendo-dropdownlist>
      </span>
      <span>{{afterMarkerValues}}</span>
    </div>
    <textarea #editor pfClickElsewhere class="formula" spellcheck="false" tabindex="-1" id="dsFormulaEditor"
      [maxlength]="formulaMaxLength"
      [attr.data-qa-id]="'txt-formula-editor'"
      [placeholder]="'Formula'"
      [(ngModel)]="inputValue"
      [disabled]="disabled ? 'true' : 'false'"
      (ngModelChange)="handleInputValueChanged($event)"
      (clickElsewhere)="handleClickElsewhere()"
      (keyup)="handleEditorKeyUp($event)"
      (keydown)="handleEditorKeyDown($event)">
    </textarea>
  </div>

  <div class="message-container">
    <div *ngIf="validating" class="validating-message">
      <fa-icon class="loading-icon" icon="circle-notch" spin="true"></fa-icon>
      <span>Validating...</span>
    </div>
  
    <div *ngIf="!validating" [style.visibility]="hideValidatingMessage ? 'hidden' : 'visible'">
      <div *ngIf="!isValid" class="validating-message text-danger">
        <fa-icon class="invalid-icon" icon="times-circle"></fa-icon>
        <span>Formula Error</span>
      </div>
    
      <div *ngIf="isValid" class="validating-message text-success">
        <fa-icon class="valid-icon" icon="check-circle"></fa-icon>
        <span>Formula Validated</span>
      </div>
    </div>
  </div>
</div>