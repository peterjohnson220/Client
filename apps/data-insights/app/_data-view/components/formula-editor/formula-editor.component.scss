@import 'data-insights-variables';

.formula-container {
  height: 150px;
  margin-bottom: 1rem;
}

.formula, .popup-anchor {
  padding: 1rem;
  font-size: 16px;
  font-weight: 300;
  position: absolute;
}

.formula {
  height: 150px;
  width: calc(100% - 2rem);
  resize: none;
  z-index: 2;
  border: 1px solid #ced4da;
  background-color: transparent;

  &:focus {
    outline-color: $background-grey;
  }
}

.popup-anchor {
  z-index: 1;
  color: transparent;
  width: 947px;
}

.suggestions-popup-content {
  min-width: 150px;
  max-height: 150px;
  overflow-y: auto;

  .suggestion {
    padding: 0.5rem 0;
    &:hover {
      background-color: #cccccc;
    }
  }
}

.validating-message {
  font-size: 14px;

  .valid-icon, .invalid-icon, .loading-icon {
    margin-right: 0.5rem;
    font-size: 16px;
  }

}
