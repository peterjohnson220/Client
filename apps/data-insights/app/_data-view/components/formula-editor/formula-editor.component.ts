import { Component, Input, ViewChild, ElementRef, Output, EventEmitter, OnInit, OnDestroy } from '@angular/core';

import { DropDownListComponent } from '@progress/kendo-angular-dropdowns';

import { SuggestionIndicator, functionNames, SuggestionIndicatorType, SpecialCharacter } from '../../models';
import { FormulaHelper } from '../../helpers';

@Component({
  selector: 'pf-formula-editor',
  templateUrl: './formula-editor.component.html',
  styleUrls: ['./formula-editor.component.scss']
})
export class FormulaEditorComponent implements OnInit {
  @Input() initialFormula: string;
  @Input() fieldSuggestions: string[];
  @Input() validating: boolean;
  @Input() isValid: boolean;
  @Input() disabled: boolean;
  @Input() isWaitingForValidation: boolean;
  @Output() formulaChanged: EventEmitter<string> = new EventEmitter<string>();

  @ViewChild('editor', { static: true }) editor: ElementRef;
  @ViewChild('suggestionList', { static: true }) suggestionList: DropDownListComponent;

  readonly MIN_QUERY_LENGTH = 1;
  readonly formulaMaxLength = 5000;
  readonly popupSettings = {
    width: 'auto',
    anchorAlign: { horizontal: 'right', vertical: 'bottom' },
    popupAlign: { horizontal: 'left', vertical: 'top' }
  };
  suggestions: string[];
  filteredSuggestions: string[];
  inputValue: string;
  searchTerm: string;
  beforeMarkerValues: string;
  markerValue: string;
  afterMarkerValues: string;
  showPopup: boolean;
  isInsertingFunctionTemplate: boolean;
  suggestionIndicator: SuggestionIndicator;

  ngOnInit(): void {
    this.initEditorData(this.initialFormula);
  }

  handleInputValueChanged(value: string): void {
    this.formulaChanged.emit(value);
    this.updatePopupMarkerValue();
    this.resetEditorIfInputValueIsEmpty();
    this.resetSuggestionIndicatorIfIndicatorRemoved();
    if (this.suggestionIndicator.Entered) {
      this.showPopupIfSuggestionsFound();
    } else {
      this.checkIfSuggestionIndicatorEntered();
    }
  }

  handleEditorKeyDown(event: any): void {
    if (event.key === 'Enter') {
      event.preventDefault();
      return;
    }
    if ((event.key === 'ArrowDown' || event.key === 'Down') && this.showPopup && !this.suggestionList.isFocused) {
      event.preventDefault();
      this.suggestionList.focus();
    }
  }

  handleEditorKeyUp(event: any): void {
    this.isInsertingFunctionTemplate = false;
  }

  handleSuggestionClicked(suggestion: string): void {
    const fromIndex: number = this.suggestionIndicator.Index;
    const toIndex: number = this.editor.nativeElement.selectionStart;
    const textToInsert = this.getFormattedSelectedSuggestion(suggestion, toIndex);
    this.insertSelectedSuggestionIntoEditor(fromIndex, toIndex, textToInsert);
    this.resetSuggestionIndicator();
    this.suggestionList.reset();
    this.setEditorFocus();
  }

  handleClickElsewhere(): void {
    this.showPopup = false;
  }

  public get hideValidatingMessage(): boolean {
    return (!this.inputValue || this.inputValue.length === 0 || this.isInsertingFunctionTemplate || this.isWaitingForValidation);
  }

  private initEditorData(formula: string): void {
    this.inputValue = formula;
    this.filteredSuggestions = [];
    this.suggestions = [];
    this.searchTerm = '';
    this.showPopup = false;
    this.suggestionIndicator = FormulaHelper.buildSuggestionIndicator();
  }

  private checkIfSuggestionIndicatorEntered(): void {
    const lastCharacterIndex: number = this.editor.nativeElement.selectionStart - 1;
    const lastCharacter = this.inputValue.charAt(lastCharacterIndex);
    switch (lastCharacter) {
      case SpecialCharacter.OpenBracket: {
        this.suggestionIndicator.Character = SpecialCharacter.OpenBracket;
        this.updateSuggestions(lastCharacterIndex, SuggestionIndicatorType.Field, this.fieldSuggestions);
        break;
      }
      case SpecialCharacter.DollarSign: {
        this.suggestionIndicator.Character = SpecialCharacter.DollarSign;
        this.updateSuggestions(lastCharacterIndex, SuggestionIndicatorType.Function, functionNames);
        break;
      }
      default: {
        break;
      }
    }
  }

  private updateSuggestions(suggestionIndicatorIndex: number, type: SuggestionIndicatorType, suggestions: string[]): void {
    this.suggestionIndicator.Entered = true;
    this.suggestionIndicator.Index = suggestionIndicatorIndex;
    this.suggestionIndicator.Type = type;
    this.suggestions = suggestions;
  }

  private resetEditorIfInputValueIsEmpty(): void {
    if (this.inputValue.length === 0) {
      this.showPopup = false;
      this.suggestionList.toggle(false);
      this.initEditorData('');
      return;
    }
  }

  private getFormattedSelectedSuggestion(suggestion: string, selectionStartIndex: number): string {
    let textToInsert = '';
    switch (this.suggestionIndicator.Type) {
      case SuggestionIndicatorType.Field: {
        const selectionStartCharacter: string = this.inputValue.charAt(selectionStartIndex);
        textToInsert = FormulaHelper.getFormattedField(suggestion, selectionStartCharacter);
        break;
      }
      case SuggestionIndicatorType.Function: {
        textToInsert = FormulaHelper.getFormattedFunction(suggestion);
        break;
      }
      default: {
        break;
      }
    }
    return textToInsert;
  }

  private resetSuggestionIndicatorIfIndicatorRemoved(): void {
    if (this.suggestionIndicator.Index === -1) {
      return;
    }
    const suggestionIndicatorRemoved: boolean = this.inputValue.charAt(this.suggestionIndicator.Index) !== this.suggestionIndicator.Character;
    if (suggestionIndicatorRemoved) {
      this.resetSuggestionIndicator();
      this.suggestionList.toggle(false);
    }
  }

  private resetSuggestionIndicator(): void {
    this.suggestionIndicator.Entered = false;
    this.suggestionIndicator.Type = null;
    this.suggestionIndicator.Index = -1;
    this.suggestionIndicator.Character = null;
    this.searchTerm = '';
    this.filteredSuggestions = [];
  }

  private insertSelectedSuggestionIntoEditor(fromIndex: number, toIndex: number, textToInsert: string): void {
    this.isInsertingFunctionTemplate = (this.suggestionIndicator.Type === SuggestionIndicatorType.Function);
    this.editor.nativeElement.focus();
    this.editor.nativeElement.setRangeText(textToInsert, fromIndex, toIndex, 'end');
    this.inputValue = this.editor.nativeElement.value;
    if (!this.isInsertingFunctionTemplate) {
      this.formulaChanged.emit(this.inputValue);
    }
  }

  private showPopupIfSuggestionsFound(): void {
    const currentPosition: number = this.editor.nativeElement.selectionStart;
    this.searchTerm = this.inputValue.substring(this.suggestionIndicator.Index + 1, currentPosition);
    if (!!this.searchTerm && this.searchTerm.length >= this.MIN_QUERY_LENGTH) {
      this.filteredSuggestions = this.suggestions.filter((s) => s.toLowerCase().indexOf(this.searchTerm.toLowerCase()) !== -1);
      this.showPopup = (this.filteredSuggestions.length !== 0);
      this.suggestionList.toggle(this.showPopup);
    } else {
      this.showPopup = false;
    }
  }

  private setEditorFocus(): void {
    setTimeout(() => {
      document.getElementById('dsFormulaEditor').focus();
    }, 100);
  }

  private updatePopupMarkerValue(): void {
    const currentPosition: number = this.editor.nativeElement.selectionStart;
    this.beforeMarkerValues = this.inputValue.substring(0, this.suggestionIndicator.Index);
    this.markerValue = this.inputValue.substring(this.suggestionIndicator.Index, currentPosition);
    this.afterMarkerValues = this.inputValue.substring(currentPosition, this.inputValue.length);
  }

}
