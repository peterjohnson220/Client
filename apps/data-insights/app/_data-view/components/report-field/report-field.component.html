<div *ngIf="field" [class.active-report-fields]="field.IsActive" class="d-flex justify-content-between field-container" (click)="toggledFieldSelection($event)">
  <div class="d-flex field-text-container" *ngIf="!isEditing" (click)="toggleEdit($event)">
    <span title="{{ field.FormulaName || field.SourceName}}" class="field-text editable">{{field.DisplayName}}</span>
  </div>
  <div class="form-group">
    <input *ngIf="isEditing"
           type="text"
           class="form-control"
           maxlength="255"
           [value]="field.DisplayName"
           (blur)="handleDisplayNameUpdated($event)"
           #inputField>
  </div>
  <div class="d-flex right-container">
    <span class="field-text" title="{{field.Entity | uppercase}}">{{field.Entity | uppercase}}</span>
    <div *ngIf="removable" class="delete-btn d-flex flex-row-reverse" (click)="handleFieldRemoved()" data-qa-id='btn-deleteElement'>
      <fa-icon icon="times"></fa-icon>
    </div>
  </div>
</div>
<!--Hidden Cell for Drag/Drop Re-Order-->
<div class="gu-mirror-show">
      selected <strong>{{activeFieldsCount}}</strong> field<ng-container *ngIf="activeFieldsCount > 1">s</ng-container> for reordering
</div>
