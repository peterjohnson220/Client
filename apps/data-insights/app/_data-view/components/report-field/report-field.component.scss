@import 'colors';
@import 'common/common-variables';
@import 'data-insights-variables';

.field-container {
  padding: 0.8rem;
  margin-top: 1rem;
  font-size: 13px;
  border-radius: 0.5rem;
  border-bottom: solid 1px $grid-border-color;
  background-color: #ffffff;

  &:hover{
    cursor: pointer;
  }
}

.field-text-container{
  max-width: 160px;

  .field-text.editable {
    cursor: default;
    padding: .35rem;
    @extend %ellipsis-overflowing-text;
  }

  .editable-icon:not(:hover) {
    display: none;
  }

  &:hover {
    .editable-icon {
      width: 20px;
      display: inline-block;
    }

    background-color: lighten($inactive-button-color, 20%);
    border-radius: 7%;
    cursor: pointer;
  }
}

.right-container {
  color: $dark-grey;

  span {
    text-align: right;
    width: 100px;
    @extend %ellipsis-overflowing-text;
  }

  .delete-btn {
    width: 20px;
    cursor: pointer;
  }

}

.active-report-fields{
  background-color: $selected-item-background-color;
  cursor: all-scroll !important;
}

.gu-mirror-show {
  padding: 0.8rem;
  margin-top: 1rem;
  font-size: 13px;
  border-radius: 0.5rem;
  border-bottom: solid 1px #d6d6d6;
  background-color: #fff;
}
