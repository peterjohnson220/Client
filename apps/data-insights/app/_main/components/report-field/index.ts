export * from './report-field.component';
