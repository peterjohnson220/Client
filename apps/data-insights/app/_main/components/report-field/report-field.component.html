<div *ngIf="field" class="d-flex justify-content-between field-container">
  <div class="d-flex field-text-container" *ngIf="!isEditing" (click)="toggleEdit()">
    <span title="{{field.SourceName}}" class="field-text editable">{{field.DisplayName}}</span>
    <span title="Click to Edit Display Name" class="editable-icon"><fa-icon icon="pencil"></fa-icon></span>
  </div>
  <div class="form-group">
    <input *ngIf="isEditing"
           type="text"
           class="form-control"
           maxlength="255"
           [value]="field.DisplayName"
           (blur)="handleDisplayNameUpdated($event, field)"
           #inputField>
  </div>
  <div class="d-flex right-container">
    <span class="field-text" title="{{field.Entity | uppercase}}">{{field.Entity | uppercase}}</span>
    <div *ngIf="removable" class="delete-btn d-flex flex-row-reverse" (click)="handleFieldRemoved()" data-qa-id='btn-deleteElement'>
      <fa-icon icon="times"></fa-icon>
    </div>
  </div>
</div>
