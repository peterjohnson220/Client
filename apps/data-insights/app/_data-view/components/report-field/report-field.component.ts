import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';

import { Field } from 'libs/features/formula-editor';

@Component({
  selector: 'pf-report-field',
  templateUrl: './report-field.component.html',
  styleUrls: ['./report-field.component.scss']
})
export class ReportFieldComponent {
  @Input() field: Field;
  @Input() removable: boolean;
  @Input() activeFieldsCount: number;
  @Output() fieldRemoved: EventEmitter<Field> = new EventEmitter();
  @Output() displayNameUpdated: EventEmitter<string> = new EventEmitter<string>();
  @Output() fieldIsActive: EventEmitter<{field: Field, multipleSelection: boolean}> = new EventEmitter<{field: Field, multipleSelection: boolean}>();

  @ViewChild('inputField') inputField: ElementRef;

  isEditing: boolean;
  multipleSelection: boolean;

  handleFieldRemoved() {
    this.fieldRemoved.emit(this.field);
  }

  toggledFieldSelection(event) {
    this.multipleSelection = !!event.ctrlKey;
    this.field.IsActive = !this.field.IsActive;
    this.fieldIsActive.emit({field: this.field, multipleSelection: this.multipleSelection});
  }

  toggleEdit(e: Event) {
    this.isEditing = !this.isEditing;
    setTimeout(() => {
      this.inputField.nativeElement.focus();
    }, 0);
    e.stopPropagation();
  }

  handleDisplayNameUpdated(event) {
    const newDisplayName = event.target.value;
    if (newDisplayName === this.field.DisplayName || newDisplayName === '') {
      this.isEditing = false;
      return;
    }

    this.displayNameUpdated.emit(newDisplayName);
    this.isEditing = false;
  }

}
