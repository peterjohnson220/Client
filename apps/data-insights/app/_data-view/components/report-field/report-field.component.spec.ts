import { ComponentFixture, TestBed } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';

import { generateMockField } from 'libs/ui/formula-editor';

import { ReportFieldComponent } from './report-field.component';


describe('Data-Insights Report Field Component', () => {
  let instance: ReportFieldComponent;
  let fixture: ComponentFixture<ReportFieldComponent>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [ ReportFieldComponent ],
      schemas: [NO_ERRORS_SCHEMA]
    });

    fixture = TestBed.createComponent(ReportFieldComponent);
    instance = fixture.componentInstance;

    fixture.detectChanges();
  });


  it('should emit the removed field when handling field removal', () => {
    jest.spyOn(instance.fieldRemoved, 'emit');
    instance.field = generateMockField();

    instance.handleFieldRemoved();

    expect(instance.fieldRemoved.emit).toHaveBeenCalledWith(instance.field);
  });

  it('should show edit form when toggleEdit is triggered', () => {
    fixture.detectChanges();
    const clickEvent = new Event('click');

    instance.toggleEdit(clickEvent);

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });
});
