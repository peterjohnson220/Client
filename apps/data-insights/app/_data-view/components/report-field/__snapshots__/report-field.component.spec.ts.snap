// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data-Insights Report Field Component should show edit form when toggleEdit is triggered 1`] = `
<pf-report-field
  displayNameUpdated={[Function EventEmitter_]}
  fieldIsActive={[Function EventEmitter_]}
  fieldRemoved={[Function EventEmitter_]}
  inputField="undefined"
  isEditing={[Function Boolean]}
>
  <div
    class="gu-mirror-show"
  >
     selected 
    <strong>
      
    </strong>
     field
    
     for reordering
  
  </div>
</pf-report-field>
`;
